src/Database/Persist/Typed.hs

```diff
@@ -70,880 +71,406 @@ import Database.Persist.Class.PersistEntity (SafeToInsert)
-            Just maxParams -> do
-                let chunkSize = maxParams `div` length (getEntityFieldsDatabase t)
-                mapM_ insertMany_' (chunksOf chunkSize vals0)
-      where
-        insertMany_' vals = do
-          conn <- ask
-          let valss = map (map toPersistValue . toPersistFields) vals
-          let sql = Text.concat
-                  [ "INSERT INTO "
-                  , connEscapeRawName conn (unEntityNameDB $ getEntityDBName t)
-                  , "("
-                  , Text.intercalate "," $ map (connEscapeRawName conn . unFieldNameDB . fieldDB) $ getEntityFieldsDatabase t
-                  , ") VALUES ("
-                  , Text.intercalate "),(" $ replicate (length valss) $ Text.intercalate "," $ map (const "?") (getEntityFieldsDatabase t)
-                  , ")"
-                  ]
-          rawExecute sql (concat valss)
-
-        t = entityDef vals0
-
-    replace k val = do
-        conn <- asks unSqlFor
-        let t = entityDef $ Just val
-        let wher = whereStmtForKey conn k
-        let sql = Text.concat
-                [ "UPDATE "
-                , connEscapeRawName conn (unEntityNameDB $ getEntityDBName t)
-                , " SET "
-                , Text.intercalate "," (map (go conn . unFieldNameDB . fieldDB) $ getEntityFieldsDatabase t)
-                , " WHERE "
-                , wher
-                ]
-            vals = map toPersistValue (toPersistFields val) `mappend` keyToValues k
-        specializeQuery $ rawExecute sql vals
-      where
-        go conn x = connEscapeRawName conn x `Text.append` "=?"
-
-    insertKey k v = specializeQuery $ insrepHelper "INSERT" [Entity k v]
-
-    repsert key value = do
-        mExisting <- get key
-        case mExisting of
-          Nothing -> insertKey key value
-          Just _  -> replace key value
-
-    delete k = do
-        conn <- asks unSqlFor
-        specializeQuery $ rawExecute (sql conn) (keyToValues k)
-      where
-        wher conn = whereStmtForKey conn k
-        sql conn = Text.concat
-            [ "DELETE FROM "
-            , connEscapeRawName conn $ unEntityNameDB $ tableDBName $ recordTypeFromKey k
-            , " WHERE "
-            , wher conn
-            ]
+instance PersistCore (SqlFor backend) where
+    newtype BackendKey (SqlFor backend) = SqlForKey {unSqlForKey :: BackendKey SqlBackend}
+        deriving
+            ( Show
+            , Read
+            , Eq
+            , Ord
+            , Num
+            , Integral
+            , PersistField
+            , PersistFieldSql
+            , PathPiece
+            , ToHttpApiData
+            , FromHttpApiData
+            , Real
+            , Enum
+            , Bounded
+            , A.ToJSON
+            , A.FromJSON
+            )
+
+downcastPersistRecordBackend :: forall record backend. PersistRecordBackend record (SqlFor backend) :- PersistRecordBackend record SqlBackend
+downcastPersistRecordBackend = unsafeCoerceConstraint
```

Comment:
`:-` is the type of entailment IIRC - so we're saying "If there's a `PersistRecordBackend record (SqlFor backend)` instance, please coerce that into `PersistRecordBackend record SqlBackend`.

This currently works because `SqlFor` is a newtype over `SqlBackend`. Can we make this safer by using the `Coercible` constraint that exists? `Coercible (SqlFor backend) SqlBackend => ...` is a redundant constraint, but at least if the `newtype` is changed then we aren't breaking things.

Reply:
Ah yea great idea